methods/utils/kd_tree.py

```diff
@@ -0,0 +1,443 @@
+        rand_state = rng.integers(0, 0xFFFF_FFFF_FFFF_FFFF, 4, np.uint64, True)
+        while len(queue) > 0:
+            pos = queue.pop()
+            d = self.ds[pos]
+            value = self.values[pos]
+            if math.isnan(value):
+                i = d
+                item: int = self.items[i] # type: ignore
+                while item != -1:
+                    # Check item
+                    found = True
+                    for d in range(self.dimensions):
+                        value = self.rows[item, d]
+                        if value < low[d]:
+                            found = False
+                            break
+                        if value > high[d]:
+                            found = False
+                            break
+                    if found:
+                        # Keep track of how many items are found for the probabilities to work out.
+                        found_count += 1
+                        # For each item we find, decide here whether to include it in the output.
+                        if len(finds) < count:
+                            # If we haven't found count items yet, we provisionaly take this item.
+                            finds.append(item)
+                        else:
+                            # Replace a random item in finds based on algorithm R.
+
+                            # When we find the N+1th item, by induction, we assume we have selected
+                            # count of the previous N items with 1/N probability.
+                            #
+                            # We want to select this item with count/N+1 probability
+                            # and we want it to replace one of the previously selected
+                            # items with a uniform 1/count probability.
+                            # We generate a uniform random number, pos, from [0, N+1)
+                            # and select this item if that number is less than count
+                            # (which corresponds to our count/N+1 probability of selection).
+                            # The position to replace the item is then given by pos,
+                            # given we now know it is a uniform random number in [0, count)
+                            # (which again matches our desired distribution).
+
+                            # Generate a random number
+                            # Source: https://prng.di.unimi.it/xoshiro256plus.c
+                            rand = rand_state[0] + rand_state[3]
+                            t = rand_state[1] << 17
+                            rand_state[2] ^= rand_state[0]
+                            rand_state[3] ^= rand_state[1]
+                            rand_state[1] ^= rand_state[2]
+                            rand_state[0] ^= rand_state[3]
+                            rand_state[2] ^= t
+                            rand_state[3] = (rand_state[3] >> 45) | (rand_state[3] << 19)
```

Comment:
Sad that we have to do this, but I see it's because of performance reasons. Can we at least pull out this code so that we're not baking into the algorithm the encryption method? The methodology does not require this particular algorithm, just we've chosen to use it for performance reasons.

Reply:
I'm not sure I quite understand: just pull it out into a function (and hope Numba inlines it), with a comment saying it is a random number but no specific algorithm is needed, and this was just chosen for speed? Or something else? (And we can only do this I suspect if Numba does inline it correctly, and we'll be passing the state around so I'm not sure it'll be particularly clearer)